File: csharp/LogicAndTrick.WikiCodeParser/Processors/TrimWhitespaceAroundBlockNodesProcessor.cs
using System.Collections.Generic;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Processors
{
    public class TrimWhitespaceAroundBlockNodesProcessor : INodeProcessor
    {
        public int Priority { get; set; } = 20;

        public bool ShouldProcess(INode node, string scope)
        {
            return node is NodeCollection;
        }

        public IEnumerable<INode> Process(Parser parser, ParseData data, INode node, string scope)
        {
            var coll = (NodeCollection) node;

            var trimStart = false;
            for (var i = 0; i < coll.Nodes.Count; i++)
            {
                var child = coll.Nodes[i];
                var next = i < coll.Nodes.Count - 1 ? coll.Nodes[i + 1] : null;
                if (child is PlainTextNode ptn)
                {
                    var text = ptn.Text;
                    if (trimStart) text = text.TrimStart();
                    if (next is HtmlNode nht && nht.IsBlockNode) text = text.TrimEnd();
                    ptn.Text = text;
                }

                child = parser.RunProcessor(child, this, data, scope);

                if (child is HtmlNode html && html.IsBlockNode)
                {
                    trimStart = true;
                    yield return UnprocessablePlainTextNode.NewLine;
                    yield return child;
                    yield return UnprocessablePlainTextNode.NewLine;
                }
                else
                {
                    trimStart = false;
                    yield return child;
                }
            }
        }
    }
}

File: php/src/Processors/TrimWhitespaceAroundBlockNodesProcessor.php
<?php

namespace LogicAndTrick\WikiCodeParser\Processors;

use LogicAndTrick\WikiCodeParser\Nodes\HtmlNode;
use LogicAndTrick\WikiCodeParser\Nodes\INode;
use LogicAndTrick\WikiCodeParser\Nodes\NodeCollection;
use LogicAndTrick\WikiCodeParser\Nodes\PlainTextNode;
use LogicAndTrick\WikiCodeParser\Nodes\UnprocessablePlainTextNode;
use LogicAndTrick\WikiCodeParser\ParseData;
use LogicAndTrick\WikiCodeParser\Parser;

class TrimWhitespaceAroundBlockNodesProcessor implements INodeProcessor
{

    /**
     * @inheritDoc
     */
    function Priority(): int
    {
        return 20;
    }

    /**
     * @inheritDoc
     */
    function ShouldProcess(INode $node, string $scope): bool
    {
        return $node instanceof NodeCollection;
    }

    /**
     * @inheritDoc
     */
    function Process(Parser $parser, ParseData $data, INode $node, string $scope): array
    {
        /* @var $coll NodeCollection */
        $coll = $node;

        $ret = [];

        $trimStart = false;
        for ($i = 0; $i < count($coll->nodes); $i++) {
            $child = $coll->nodes[$i];
            $next = $i < count($coll->nodes) - 1 ? $coll->nodes[$i + 1] : null;
            if ($child instanceof PlainTextNode) {
                $text = $child->text;
                if ($trimStart) $text = ltrim($text);
                if ($next instanceof HtmlNode && $next->isBlockNode) $text = rtrim($text);
                $child->text = $text;
            }

            $child = $parser->RunProcessor($child, $this, $data, $scope);

            if ($child instanceof HtmlNode && $child->isBlockNode) {
                $trimStart = true;
                $ret[] = UnprocessablePlainTextNode::NewLine();
                $ret[] = $child;
                $ret[] = UnprocessablePlainTextNode::NewLine();
            } else {
                $trimStart = false;
                $ret[] = $child;
            }
        }

        return $ret;
    }
}